### PR TITLE
Added missing tests for rclcpp lifecycle

### PR DESCRIPTION
**Pull request by ahcorde**
- remove_on_set_parameters_callback
 - notify_graph_change
 - get_service_names_and_types_by_node

Related to this https://ci.ros2.org/job/nightly_linux_coverage/lastBuild/cobertura/src_ros2_rclcpp_rclcpp_lifecycle_src/lifecycle_node_cpp/lifecycle_node_cpp/

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11628/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6760/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9514/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11604/)

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 1]
```diff
@@ -355,3 +356,55 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
...
+TEST_F(TestLifecycleServiceClient, get_service_names_and_types_by_node)
+{
+  auto node1 = std::make_shared<LifecycleServiceClient>("client1");
+  auto node2 = std::make_shared<LifecycleServiceClient>("client2");
+
+  const auto * node_graph =
+    dynamic_cast<rclcpp::node_interfaces::NodeGraph *>(node1->get_node_graph_interface().get());
+  ASSERT_NE(nullptr, node_graph);
+
+  EXPECT_THROW(
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:370]: You can be the first user of my new testing macros! I've added `RCLCPP_EXPECT_THROW_EQ` and `RCLCPP_ASSERT_THROW_EQ`, in https://github.com/ros2/rclcpp/blob/master/rclcpp/test/utils/rclcpp_gtest_macros.hpp.

It has the benefit of not only checking that this throws a runtime error, but that the message contained in the runtime error matches the expected. It's also implemented for RCLError types. Obviously, only use these versions when the contents of the exception can be depended on.

See example uses here:
https://github.com/ros2/rclcpp/blob/master/rclcpp/test/utils/test_rclcpp_gtest_macros.cpp#L42-L56

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 2]
```diff
@@ -355,3 +356,55 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
...
+  auto service_names_and_types2 = node_graph->get_service_names_and_types_by_node("client2", "/");
+  EXPECT_EQ(service_names_and_types1.size(), service_names_and_types2.size());
+}
+
+TEST_F(TestLifecycleServiceClient, declare_parameter_with_no_initial_values)
+{
+  auto node1 = std::make_shared<LifecycleServiceClient>("client1");
+
+  auto on_set_parameters =
+    [](const std::vector<rclcpp::Parameter> & parameters) {
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:383]: nit: if you don't need to use `parameters` in this lambda, you can just omit the name of the argument so you don't have to cast to void.
```suggestion
    [](const std::vector<rclcpp::Parameter> &) {
```

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 3]
```diff
@@ -355,3 +356,55 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
...
+
+  auto handler = node1->add_on_set_parameters_callback(on_set_parameters);
+  RCLCPP_SCOPE_EXIT({node1->remove_on_set_parameters_callback(handler.get());});    // always reset
+}
+
+TEST_F(TestLifecycleServiceClient, wait_for_graph_change)
+{
+  auto node = std::make_shared<LifecycleServiceClient>("client_wait_for_graph_change");
+  auto * node_graph =
+    dynamic_cast<rclcpp::node_interfaces::NodeGraph *>(node->get_node_graph_interface().get());
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:398]: Is the dynamic_cast necessary?

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp (hunk 1 above)
Comment (ahcorde) [reply rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:370]: This test is in `rclcpp_lifecycle`, I can't include the header `rclcpp_gtest_macros.hpp`. Did you expose this macro/header to other packages ?

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp (hunk 3 above)
Comment (ahcorde) [reply rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:398]: I copy this line from the original test 
https://github.com/ros2/rclcpp/blob/7e68d3549cb4162630446a1cadf62a4b2f64cc82/rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp#L47-L48

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp (hunk 1 above)
Comment (brawner) [reply rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:370]: Oh you're right. Oh well, we can wait until it becomes more mainsteam.

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp (hunk 3 above)
Comment (brawner) [reply rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:397]: That test was for node_interfaces explicitly, and I wanted to verify that the cast to NodeGraph* was doing what I thought it was.

I think for this test, the std::shared_ptr<NodeGraphInterface> should be enough.

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 4]
```diff
@@ -355,3 +356,53 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
...
   EXPECT_EQ(transitions.size(), 0u);
 }
+
+
+TEST_F(TestLifecycleServiceClient, get_service_names_and_types_by_node)
+{
+  auto node1 = std::make_shared<LifecycleServiceClient>("client1");
+  auto node2 = std::make_shared<LifecycleServiceClient>("client2");
+
+  const auto * node_graph =
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:366]: Also no need for dynamic_cast here.
```suggestion
  const auto node_graph = node1->get_node_graph_interface();
```

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 5]
```diff
@@ -355,3 +356,53 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
...
+    };
+
+  auto handler = node1->add_on_set_parameters_callback(on_set_parameters);
+  RCLCPP_SCOPE_EXIT({node1->remove_on_set_parameters_callback(handler.get());});    // always reset
+}
+
+TEST_F(TestLifecycleServiceClient, wait_for_graph_change)
+{
+  auto node = std::make_shared<LifecycleServiceClient>("client_wait_for_graph_change");
+  std::shared_ptr<rclcpp::node_interfaces::NodeGraphInterface> node_graph = node->get_node_graph_interface();
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:396]:
```suggestion
  auto node_graph = node->get_node_graph_interface();
```

Feel free to use auto here. You can also leave it if you would like.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11671/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6788/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9539/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11638/)

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 6]
```diff
@@ -355,3 +356,52 @@ TEST_F(TestLifecycleServiceClient, lifecycle_transitions) {
   transitions = lifecycle_client()->get_available_transitions();
   EXPECT_EQ(transitions.size(), 0u);
 }
+
+
```
Comment (brawner) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:360]: Remove extra line

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11900/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6940/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9696/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11832/)

**Pull request merged**